### ISSUE TITLE
Albums::get times out on big gallery

### ISSUE THREAD

**Issue by matthijskooijman** [LycheeOrg/Lychee#316]
Albums::get (and thus the main gallery view) times out when there are a lot of albums and/or photos.

I started uploading a lot of older albums to Lychee recently, and now I can no longer see the album list when logged in. When not logged in, the gallery does load quickly (but shows no albums, since all are private still).

The problematic gallery has 232 albums (nested a few levels deep) with 8701 photographs in total.

The failing URL is `/api/Albums::get`, which returns 500. The error log says:

```
PHP Fatal error:  Maximum execution time of 30 seconds exceeded in /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Application.php on line 753
PHP Stack trace:
PHP   1. {main}() /Lychee-Laravel/public/index.php:0
PHP   2. App\Http\Kernel->handle() /Lychee-Laravel/public/index.php:53
PHP   3. App\Http\Kernel->sendRequestThroughRouter() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:116
PHP   4. Illuminate\Routing\Pipeline->then() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:151
PHP   5. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:104
PHP   6. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP   7. Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP   8. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php:62
PHP   9. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  10. Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  11. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php:27
PHP  12. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  13. App\Http\Middleware\TrimStrings->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  14. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21
PHP  15. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  16. Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  17. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21
PHP  18. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  19. App\Http\Middleware\TrustProxies->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  20. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/fideloper/proxy/src/TrustProxies.php:57
PHP  21. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  22. Bepsvpt\SecureHeaders\SecureHeadersMiddleware->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  23. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:28-36}() /Lychee-Laravel/vendor/bepsvpt/secure-headers/src/SecureHeadersMiddleware.php:21
PHP  24. App\Http\Kernel->Illuminate\Foundation\Http\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:173-177}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:30
PHP  25. Illuminate\Routing\Router->dispatch() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:176
PHP  26. Illuminate\Routing\Router->dispatchToRoute() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php:612
PHP  27. Illuminate\Routing\Router->runRoute() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php:623
PHP  28. Illuminate\Routing\Router->runRouteWithinStack() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php:657
PHP  29. Illuminate\Routing\Pipeline->then() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php:682
PHP  30. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:104
PHP  31. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  32. App\Http\Middleware\EncryptCookies->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  33. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php:66
PHP  34. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  35. Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  36. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php:37
PHP  37. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  38. Illuminate\Session\Middleware\StartSession->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  39. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php:56
PHP  40. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  41. Illuminate\View\Middleware\ShareErrorsFromSession->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  42. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php:49
PHP  43. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  44. Illuminate\Session\Middleware\AuthenticateSession->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  45. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Session/Middleware/AuthenticateSession.php:39
PHP  46. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  47. App\Http\Middleware\VerifyCsrfToken->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  48. App\Http\Middleware\VerifyCsrfToken->handle() /Lychee-Laravel/app/Http/Middleware/VerifyCsrfToken.php:60
PHP  49. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:47-59}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php:75
PHP  50. Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:140-169}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:53
PHP  51. Illuminate\Routing\Middleware\SubstituteBindings->handle() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:163
PHP  52. Illuminate\Routing\Pipeline->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:28-36}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php:41
PHP  53. Illuminate\Routing\Router->Illuminate\Routing\{closure:/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php:678-682}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php:30
PHP  54. Illuminate\Routing\Route->run() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php:680
PHP  55. Illuminate\Routing\Route->runController() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Route.php:176
PHP  56. Illuminate\Routing\ControllerDispatcher->dispatch() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Route.php:219
PHP  57. App\Http\Controllers\AlbumsController->callAction() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php:45
PHP  58. call_user_func_array:{/Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Controller.php:54}() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Controller.php:54
PHP  59. App\Http\Controllers\AlbumsController->get() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Routing/Controller.php:54
PHP  60. App\ModelFunctions\AlbumFunctions->prepare_albums() /Lychee-Laravel/app/Http/Controllers/AlbumsController.php:55
PHP  61. App\ModelFunctions\AlbumFunctions->get_albums() /Lychee-Laravel/app/ModelFunctions/AlbumFunctions.php:245
PHP  62. App\ModelFunctions\AlbumFunctions->get_sub_albums() /Lychee-Laravel/app/ModelFunctions/AlbumFunctions.php:446
PHP  63. App\ModelFunctions\AlbumFunctions->get_sub_albums() /Lychee-Laravel/app/ModelFunctions/AlbumFunctions.php:516
PHP  64. App\ModelFunctions\AlbumFunctions->get_sub_albums() /Lychee-Laravel/app/ModelFunctions/AlbumFunctions.php:516
PHP  65. App\ControllerFunctions\ReadAccessFunctions->album() /Lychee-Laravel/app/ModelFunctions/AlbumFunctions.php:510
PHP  66. App\ModelFunctions\SessionFunctions->is_current_user() /Lychee-Laravel/app/ControllerFunctions/ReadAccessFunctions.php:68
PHP  67. Illuminate\Support\Facades\Facade::get() /Lychee-Laravel/app/ModelFunctions/SessionFunctions.php:70
PHP  68. Illuminate\Support\Facades\Facade::__callStatic() /Lychee-Laravel/app/ModelFunctions/SessionFunctions.php:70
PHP  69. Illuminate\Session\SessionManager->get() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php:237
PHP  70. Illuminate\Session\SessionManager->__call() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php:237
PHP  71. Illuminate\Session\SessionManager->driver() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Support/Manager.php:146
PHP  72. Illuminate\Session\SessionManager->getDefaultDriver() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Support/Manager.php:59
PHP  73. Illuminate\Foundation\Application->offsetGet() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Session/SessionManager.php:210
PHP  74. Illuminate\Foundation\Application->make() /Lychee-Laravel/vendor/laravel/framework/src/Illuminate/Container/Container.php:1233
```

### Output of the diagnostics
I can't access the settings right now, since the mainpage fails to load...

**Comment by ildyria** [comment]
> I can't access the settings right now, since the mainpage fails to load...

If you are able to log in then you can access : https://example.com/Diagnostics

**Comment by d7415** [comment]
This doesn't look good, but please check the php.ini settings [here](https://github.com/LycheeOrg/Lychee-Laravel/wiki/Install#72-php). Your `max_execution_time` seems to be set to 30s.

**Comment by matthijskooijman** [comment]
@ildyria, thanks for that link, diagnostic output is included below. I do not have the symlink feature enabled, so I do not think that is the problem.

```
    Diagnostics
    -----------
    Warning: Dropbox import not working. dropbox_key is empty.
    Warning: You may experience problems when uploading a large amount of photos. Take a look in the FAQ for details.

    System Information
    ------------------
    Lychee-front Version:      3.2.16
    Lychee Version (git):      39a7e7e (master) - Probably more than 30 commits behind master
    DB Version:                040000
    System:                    Linux
    PHP Version:               7.3
    MySQL Version:             10.1.26-MariaDB-0+deb9u1
    
    Lychee total space:        23.86 GB
    Upload folder space:       23.81 GB
    System total space:        39.25 GB
    System free space:         13.48 GB (34%)
    
    Imagick:                   1
    Imagick Active:            1
    Imagick Version:           1690
    GD Version:                2.2.5

    Config Information
    ------------------
    version:                   040000
    check_for_updates:         1
    sorting_Photos_col:        takestamp
    sorting_Photos_order:      ASC
    sorting_Albums_col:        description
    sorting_Albums_order:      DESC
    imagick:                   1
    skip_duplicates:           0
    small_max_width:           0
    small_max_height:          360
    medium_max_width:          1920
    medium_max_height:         1080
    lang:                      en
    layout:                    1
    image_overlay:             0
    image_overlay_type:        exif
    default_license:           reserved
    compression_quality:       90
    full_photo:                1
    delete_imported:           1
    Mod_Frame:                 0
    Mod_Frame_refresh:         30000
    landing_page_enable:       0
    landing_owner:             John Smith
    landing_title:             John Smith
    landing_subtitle:          Cat, Dogs & Humans Photography
    landing_facebook:          https://www.facebook.com/JohnSmith
    landing_flickr:            https://www.flickr.com/JohnSmith
    landing_twitter:           https://www.twitter.com/JohnSmith
    landing_instagram:         https://instagram.com/JohnSmith
    landing_youtube:           https://www.youtube.com/JohnSmith
    landing_background:        dist/cat.jpg
    thumb_2x:                  1
    small_2x:                  0
    medium_2x:                 0
    site_title:                Lychee v4
    site_copyright_enable:     0
    site_copyright_begin:      2019
    site_copyright_end:        2019
    public_search:             0
    additional_footer_text:    
    display_social_in_gallery: 0
    public_recent:             0
    recent_age:                1
    public_starred:            0
    downloadable:              0
    photos_wraparound:         1
```

> This doesn't look good, but please check the php.ini settings here. Your max_execution_time seems to be set to 30s.

@d7415, but 30 seconds should be more than enough to show the main page of the gallery? Or are you suggesting that there is some one-time processing that needs to be done when an album of photo is first displayed? Shouldn't that have been done at import time already?

I just tried increasing the max execution time to 200s, which indeed allows viewing the main page, but the request still takes 38s every time, which is not really usable. So, as mentioned, some optimization is needed here :-)

**Comment by ildyria** [comment]
Can you check in your `.env` that:
```
APP_DEBUG=true
```
is set ... to `false` :sweat_smile:

**Comment by matthijskooijman** [comment]
Yup, debug is set to false:

```
$ cat .env |grep DEBUG
APP_DEBUG=false
```

**Comment by kamil4** [comment]
As @d7415 said, if `Albums::get` takes 38s for you then we have no good news right now. Lychee as currently implemented will be effectively unusable for you :frowning_face:.

`Albums::get` doesn't just return the top-level albums; it returns the whole album hierarchy. And for each of those albums, including the nested ones, it needs to calculate which three pictures inside are to be the thumbs of that album (we don't do that ahead of time because the result may depend on who is making the query). It gets worse. Nearly every time we make edits to the gallery in the front end (upload new pictures, move things around, etc.), the cached result of `Albums::get` is dropped and we request an updated copy from the server -- that will be another 38s for you, which is why I said that it won't be usable in practice.

Why is it done that way? Well, I guess we never imagined that `Albums::get` could take more than a second or two. There are people with way more than 8701 pictures who don't have this problem so I'm guessing it must be the 232 nested albums that kill the performance for you.

Can you give us some more details? Are you logged in as administrator when this happens? Do the albums and photos belong to the administrator or did you create a separate user for them? If the albums are public, what is the performance like when accessing the gallery without logging in (I wouldn't be surprised if it was worse but I would like to have that confirmed).

How nested are those albums? Perhaps you could share with us the output of the following query run in the database console:

```
select id, parent_id, owner_id from albums;
```

Are the photos distributed relatively uniformly among albums or are there a few massive albums and the rest are much smaller? I would ask you to give us the output of `select album_id from photos;` but that would be >8000 lines...

**Comment by kamil4** [comment]
@matthijskooijman For the second query, could you redirect the output to a file, like so:
```
mysql --user <user> --password -e 'use <database_name>; select album_id from photos;' >mysql.out
```
And then in the shell do:
```
sort mysql.out | uniq -c
```
That should make the output much more reasonable while giving us the information we may need...

**Comment by ildyria** [comment]
The thumbnail generation may be one of the reason. :x

**Comment by kamil4** [comment]
Yes, that is my feeling as well. For nested albums, the complexity is probably at least quadratic. I think we can optimize it by ensuring that we go recursively depth-first and then for the parent albums we don't query the child album content again, but simply take the already calculated thumbs from them and merge them in the query with their siblings and the content of the parent album. This would also cut down on the number of access checks that we currently do.

However, the scientist inside me is yelling that I should reproduce the problem first before blindly hacking on the code :wink:. Hence my request for more data.

**Comment by ildyria** [comment]
you should request the album table too then.

**Comment by kamil4** [comment]
Erm, I thought I did?
> ```
> select id, parent_id, owner_id from albums;
> ```

**Comment by d7415** [comment]
> 
> ```
> mysql --user <user> --password -e 'use <database_name>; select album_id from photos;' >mysql.out
> ```
I'm not at my PC, but...

`select album_id, count(album_ID) from photos group by album_ID;`?

**Comment by kamil4** [comment]
> I'm not at my PC, but...
> 
> `select album_id, count(album_ID) from photos group by album_ID;`?

It works! What can I say, I'm an ignorant when it comes to SQL. :smiley:

**Comment by matthijskooijman** [comment]
Woah, lots of response. Thanks for that! I'll try to provide some extra info.

> Are you logged in as administrator when this happens?

Yes

> Do the albums and photos belong to the administrator or did you create a separate user for them?

There is just one user, which imported (not uploaded) all the photos.

> If the albums are public, what is the performance like when accessing the gallery without logging in (I wouldn't be surprised if it was worse but I would like to have that confirmed).

None of the albums are public, until I increased the max execution time I haven't been able to make any of them public. I can try making them public through the database (which was already my plan), if that helps?

> How nested are those albums?

I think they are pretty balanced. One top level album with mostly 3-5 sublevels of albums. Most albums have 4-8 subalbums.

> The thumbnail generation may be one of the reason. :x
Aren't thumbnails generated on upload/import? Also, I'm seeing this 38s request time on repeated requests, while I expect thumbnails are generated only once?

> For nested albums, the complexity is probably at least quadratic. I think we can optimize it by ensuring that we go recursively depth-first and then for the parent albums we don't query the child album content again, but simply take the already calculated thumbs from them and merge them in the query with their siblings and the content of the parent album. 

Ah, I can see how that would end up killing performance in this case. If I am thinking about this properly, I guess this would end up processing each photograph a number of times equal to its nesting depth, so probably 4-6 times on average, so fixing that would already be a big gain (but perhaps not all of it, yet..).

For completeness, here's two more remarks:
 - I'm running on fairly slow hardware (an older Intel Atom server), so any performance problems are probably emphasized in this setup.
 - The import of the photos (from a local server dir) did not complete in one go. I've restarted it once (after a 1500 photos or so) and then it stopped again, I haven't restarted it to complete the import yet. I think it should not matter here, but in case it does, I wanted to mention it.

> Perhaps you could share with us the output of the following query run in the database console:

I've taken the liberty to integrate your queries into one :-)

```
select id, parent_id, owner_id, (select count(id) from photos where album_id = albums.id) as photo_count from albums;
id	parent_id	owner_id	photo_count
15647683656423	NULL	0	0
15647684721376	15647683656423	0	0
15647684722208	15647684721376	0	0
15647684724353	15647684722208	0	44
15647685380511	15647684722208	0	19
15647685720707	15647684721376	0	0
15647685721981	15647685720707	0	0
15647685722749	15647685721981	0	121
15647693151817	15647685720707	0	0
15647693153331	15647693151817	0	68
15647700154349	15647685720707	0	0
15647700155661	15647700154349	0	86
15647707495230	15647685720707	0	0
15647707496746	15647707495230	0	0
15647707497660	15647707496746	0	0
15647707498617	15647707497660	0	23
15647710169805	15647684721376	0	0
15647710171538	15647710169805	0	0
15647710172398	15647710171538	0	7
15647710264840	15647710169805	0	0
15647710266345	15647710264840	0	0
15647710267221	15647710266345	0	0
15647710268105	15647710267221	0	30
15647713100544	15647710266345	0	0
15647713102278	15647713100544	0	28
15647715944064	15647710266345	0	0
15647715945657	15647715944064	0	75
15647722639644	15647710264840	0	0
15647722641248	15647722639644	0	0
15647722642138	15647722641248	0	17
15647724085159	15647722639644	0	0
15647724086763	15647724085159	0	21
15647724623193	15647684721376	0	0
15647724624912	15647724623193	0	0
15647724625723	15647724624912	0	14
15647725586514	15647724624912	0	46
15647737448928	15647724624912	0	13
15647737631903	15647724624912	0	22
15647738633065	15647724624912	0	96
15647748704321	15647724623193	0	0
15647748705843	15647748704321	0	22
15647748941378	15647748704321	0	17
15647749223803	15647748704321	0	0
15647749225356	15647749223803	0	59
15647752303143	15647748704321	0	184
15647758324631	15647748704321	0	0
15647758326377	15647758324631	0	52
15647761381295	15647748704321	0	0
15647761382850	15647761381295	0	82
15647765352987	15647724623193	0	0
15647765354721	15647765352987	0	187
15647779851261	15647765352987	0	121
15647786100138	15647765352987	0	45
15647788260576	15647765352987	0	72
15647791693420	15647724623193	0	0
15647791694937	15647791693420	0	73
15647796714272	15647791693420	0	46
15647800231800	15647791693420	0	58
15647801237764	15647791693420	0	64
15647806024667	15647791693420	0	124
15647816544794	15647724623193	0	0
15647816546419	15647816544794	0	32
15647817698597	15647684721376	0	0
15647817700176	15647817698597	0	0
15647817701129	15647817700176	0	198
15647819015380	15647816544794	0	69
15647822953788	15647816544794	0	88
15647830952274	15647816544794	0	299
15647843153538	15647684721376	0	0
15647843156021	15647843153538	0	0
15647843157345	15647843156021	0	0
15647843158622	15647843157345	0	131
15647859762114	15647843153538	0	0
15647859765830	15647859762114	0	92
15647867534137	15647724623193	0	0
15647867538174	15647867534137	0	69
15647867587904	15647843153538	0	0
15647867592252	15647867587904	0	82
15647869362862	15647867534137	0	74
15647870833666	15647867534137	0	0
15647870837709	15647870833666	0	29
15647873213689	15647867587904	0	31
15647874732545	15647867534137	0	0
15647874735780	15647874732545	0	108
15647876554094	15647867587904	0	112
15647887934430	15647867587904	0	52
15647891006778	15647867534137	0	0
15647891008958	15647891006778	0	187
15647891239743	15647843153538	0	0
15647891241757	15647891239743	0	22
15647892672354	15647891239743	0	108
15647906789680	15647891239743	0	0
15647906792575	15647906789680	0	0
15647906793598	15647906792575	0	0
15647906794873	15647906793598	0	48
15647912767733	15647724623193	0	0
15647912773803	15647912767733	0	49
15647914057993	15647843153538	0	0
15647914060460	15647914057993	0	16
15647914350698	15647914057993	0	55
15647918290005	15647914057993	0	296
15647919687134	15647912767733	0	22
15647921233653	15647912767733	0	107
15647923415414	15647912767733	0	66
15647933170805	15647724623193	0	0
15647933173571	15647933170805	0	40
15647938993321	15647933170805	0	0
15647938996985	15647938993321	0	7
15647939102122	15647938993321	0	90
15647940939893	15647938993321	0	141
15647943514688	15647933170805	0	57
15647948951289	15647843153538	0	0
15647948953012	15647948951289	0	240
15647950647769	15647933170805	0	117
15647966753130	15647948951289	0	24
15647968158641	15647948951289	0	244
15647984776758	15647843153538	0	0
15647984778504	15647984776758	0	44
15647988557203	15647984776758	0	164
15648000972883	15647984776758	0	7
15648001531176	15647843153538	0	0
15648001532114	15648001531176	0	69
15648006405419	15648001531176	0	11
15648007041892	15648001531176	0	5
15648007420423	15648001531176	0	97
15648015523441	15647843153538	0	0
15648015525067	15648015523441	0	0
15648015525830	15648015525067	0	132
15648027337504	15648015523441	0	0
15648027339322	15648027337504	0	124
15648037548928	15647843153538	0	0
15648037550463	15648037548928	0	0
15648037551516	15648037550463	0	72
15648041085623	15648037548928	0	0
15648041087691	15648041085623	0	42
15648045371307	15648037548928	0	0
15648045373150	15648045371307	0	0
15648045373908	15648045373150	0	108
15648056835941	15648045373150	0	30
15648059775064	15648045373150	0	40
15648062867276	15648045373150	0	15
15648064426048	15648037548928	0	0
15648064427829	15648064426048	0	75
15648070642751	15647843153538	0	0
15648070644300	15648070642751	0	63
15648071730816	15648070644300	0	61
15648071969696	15648070642751	0	82
15648073019504	15648070642751	0	22
15648077357452	15647843153538	0	0
15648077358904	15648077357452	0	201
15648081101220	15648077357452	0	120
15648082561109	15648077357452	0	0
15648082561897	15648082561109	0	0
15648082562674	15648082561897	0	0
15648082563460	15648082562674	0	22
15648085211593	15647843153538	0	0
15648085213219	15648085211593	0	77
15648091852190	15647684721376	0	0
15648091853877	15648091852190	0	0
15648091854677	15648091853877	0	32
15648095022998	15648091852190	0	0
15648095024506	15648095022998	0	0
15648095025263	15648095024506	0	96
15648096929529	15648095022998	0	0
15648096931003	15648096929529	0	13
15648098583386	15648095022998	0	0
15648098584901	15648098583386	0	9
15648099271319	15648091852190	0	0
15648099273092	15648099271319	0	0
15648099273918	15648099273092	0	62
15648109049920	15647684721376	0	0
15648109051436	15648109049920	0	0
15648109052302	15648109051436	0	0
15648109053189	15648109052302	0	22
15648111951665	15648109051436	0	0
15648111953386	15648111951665	0	24
15648114295814	15648109051436	0	0
15648114297462	15648114295814	0	108
15648129505990	15647684721376	0	0
15648129507499	15648129505990	0	0
15648129508349	15648129507499	0	7
15648130043755	15648129507499	0	0
15648130045277	15648130043755	0	115
15648141234569	15647684721376	0	0
15648141236314	15648141234569	0	0
15648141237183	15648141236314	0	14
15648141930951	15648141236314	0	1
15648141948744	15648141930951	0	44
15648142610988	15648141236314	0	7
15648142941076	15648141236314	0	6
15648143031853	15648141236314	0	4
15648143175192	15648141234569	0	0
15648143176017	15648143175192	0	0
15648143176901	15648143176017	0	47
15648147475373	15647684721376	0	0
15648147476933	15648147475373	0	0
15648147477838	15648147476933	0	134
15648160036873	15647684721376	0	0
15648160038207	15648160036873	0	0
15648160039184	15648160038207	0	0
15648160039964	15648160039184	0	48
15648165598762	15648160038207	0	0
15648165600273	15648165598762	0	185
15648211188801	15647684721376	0	0
15648211192203	15648211188801	0	0
15648211193545	15648211192203	0	0
15648211195115	15648211193545	0	48
15648212537001	15648211192203	0	0
15648212539227	15648212537001	0	6
15648213159610	15648211192203	0	62
15648213777245	15648213159610	0	62
15648214248381	15648211192203	0	0
15648214250176	15648214248381	0	88
15648224551162	15648211188801	0	0
15648224553522	15648224551162	0	0
15648224554406	15648224553522	0	77
15648231056108	15648224551162	0	0
15648231057730	15648231056108	0	17
15648233266533	15648224551162	0	0
15648233268298	15648233266533	0	27
15648236284608	15648211188801	0	0
15648236285790	15648236284608	0	0
15648236286913	15648236285790	0	62
15648244683582	15648236284608	0	0
15648244685649	15648244683582	0	108
15648255905466	15648211188801	0	0
15648255907089	15648255905466	0	0
15648255907966	15648255907089	0	9
15648257071178	15648255905466	0	0
15648257072794	15648257071178	0	36
15648260266471	15648255905466	0	0
15648260268191	15648260266471	0	100
```

And, in case it helps for reproducing, here's my complete database (I've only omitted the log table content and the username & password from the config table, but that should be no problem for you :-)

[db_sql.txt](https://github.com/LycheeOrg/Lychee-Laravel/files/3502745/db_sql.txt)

**Comment by kamil4** [comment]
> Woah, lots of response. Thanks for that!

Well, I guess our ears perk up when we get a report of performance problems on the server side. If the problem was in the front end, it would probably be a different story :smile:.

> None of the albums are public, until I increased the max execution time I haven't been able to make any of them public. I can try making them public through the database (which was already my plan), if that helps?

Nah, don't bother. Most likely it will perform worse then (if the user is the administrator, we skip some access checks).

> > The thumbnail generation may be one of the reason. :x
> 
> Aren't thumbnails generated on upload/import? Also, I'm seeing this 38s request time on repeated requests, while I expect thumbnails are generated only once?

Thumbnails are indeed generated on upload/import. What @ildyria was referring to wasn't the generation of images but rather the determination of which three out of hundreds, possibly thousands of images inside the album to display on the album cover. For a flat album it's a single SQL query but for nested ones...

> > For nested albums, the complexity is probably at least quadratic. I think we can optimize it by ensuring that we go recursively depth-first and then for the parent albums we don't query the child album content again, but simply take the already calculated thumbs from them and merge them in the query with their siblings and the content of the parent album.
> 
> Ah, I can see how that would end up killing performance in this case. If I am thinking about this properly, I guess this would end up processing each photograph a number of times equal to its nesting depth, so probably 4-6 times on average, so fixing that would already be a big gain (but perhaps not all of it, yet..).

While we don't reprocess the images themselves, we issue a lot of SQL queries to build a list of nested subalbums inside each album and to check access permissions (we need to consider, for instance, non-public subalbums inside a public album). We currently do it completely independently at each nested level of the hierarchy, which is clearly suboptimal (we should be doing it at the deepest levels first and then reuse the results from deeper levels when calculating at higher levels). It remains to be seen whether this is the only (or even the main) problem we have but it is the low-hanging fruit...

> And, in case it helps for reproducing, here's my complete database (I've only omitted the log table content and the username & password from the config table, but that should be no problem for you :-)

Thanks, that should make things easier for us! We've been all rather busy lately with our dayjobs and such so I can't make you any promises but as I said, problems like this do tend to draw our attention more than some others...

**Comment by kamil4** [comment]
I'm having difficulties reproducing this problem even with your database. When I reload at the top level (the albums view) after logging in, `Albums::get` takes <800ms for me. Granted, I recently updated my VPS to a 3GHz+ Skylake core and NVMe storage, but still, it shouldn't be nearly 50x faster...

**Comment by matthijskooijman** [comment]
> What @ildyria was referring to wasn't the generation of images but rather the determination of which three out of hundreds, possibly thousands of images inside the album to display on the album cover.

Ah, of course :-)

> While we don't reprocess the images themselves, we issue a lot of SQL queries to build a list of nested subalbums inside each album and to check access permissions (we need to consider, for instance, non-public subalbums inside a public album)

Yeah, that's what I meant with "processing", just using a single photograph as the "unit of work", in the complexity sense, i.e. O(number_of_photos * depth)

> We've been all rather busy lately with our dayjobs and such so I can't make you any promises but as I said, problems like this do tend to draw our attention more than some others...

I was planning to fix it myself, but then my dayjob got in the way, so I thought to at least post an issue of it :-p So I can totally relate, no worries :-)

> I'm having difficulties reproducing this problem even with your database. When I reload at the top level (the albums view) after logging in, Albums::get takes <800ms for me. Granted, I recently updated my VPS to a 3GHz+ Skylake core and NVMe storage, but still, it shouldn't be nearly 50x faster...

Well, my server *is* quite slow. I looked up some CPU benchmarks for comparison:

https://www.cpubenchmark.net/compare/Intel-i7-620M-vs-Intel-i5-6267U-vs-Intel-Atom-D510/849vs2757vs610

The Atom D510 is the server that takes 38s, the 620M is my laptop (where I just tried and where it needs 2.4s). So based on the raw CPU benchmark score, you would expect less slowdown, but perhaps this particular workload is somehow more sensitive to slowdown on the Atom architecture, which has less predictive execution, probably less cache, an older and slower memory bus, etc.

So that does mean the problem is not as big in general as it seemed (though it still is for me :-p), but 800ms on high-end hardware might still be problematic on a busy site, so optimization still seems relevant.

**Comment by ildyria** [comment]
> (I've only omitted the log table content and the username & password from the config table, but that should be no problem for you :-)

Just wanted to point out that both your username and password are bcrypt hashed... so you would still have been pretty safe in case of a leak. :)

**Comment by ildyria** [comment]
> So that does mean the problem is not as big in general as it seemed (though it still is for me :-p), but 800ms on high-end hardware might still be problematic on a busy site, so optimization still seems relevant.

The problem is we can't really cache the thumbnail as they are decided "on the fly" given whether the user is logged, whether the sub album is public or not, whether the album is shared with another user etc...

Caching would imply to create a pivot table between `album_id` - `user_id` - `photo_id` which would be updated on `upload` / `move` / `delete` / `merge` for each cases... Sound pretty heavy implementation wise.

**Comment by matthijskooijman** [comment]
> Just wanted to point out that both your username and password are bcrypt hashed... so you would still have been pretty safe in case of a leak. :)

Yeah, and it's just a random, unique password anyway, but well :-)

> The problem is we can't really cache the thumbnail

Yeah, I can see that. But the modification to the depth-first-and-reuse-results structure proposed by @kamil4 would be possible and likely help a lot already.

**Comment by kamil4** [comment]
@matthijskooijman Can you try branch `fix-316` from my PR #318, which implements the optimization I mentioned? It cuts down the execution time on my system to 400ms; I'm curious what it will do on yours...

**Comment by matthijskooijman** [comment]
I just tried it, time is down to 22s now. That's less than I expected, so maybe there is another code path, completely separate from this recursion, that is also taking a lot of time?

I also noticed that when I enter subalbums, it runs Album::get (rather than Albums::get), which is again slow (since I have a single top-level album, entering that will take another 22s).

I'm away today, but I can try to add a bit of instrumentation to the code this weekend, to confirm what the time is spent on. One obvious thing to check is the SQL queries that are made (maybe some of them can be combined, or maybe indexes can be added to speed up some of them).

**Comment by kamil4** [comment]
Well, it nearly halved the time for you, which is in line with what I saw on my system. Not bad for a 20-line patch but yeah, not good enough yet to make Lychee usable on your system.

I'm mildly curious if the bottleneck on your system is the CPU or storage. If you care to investigate, maybe try moving the database storage to tmpfs and see what that does performance-wise? Your system is not running out of RAM and swapping out while handling the queries, is it?

`Album::get` performs most of the same actions as `Albums::get`, only starting from the current album rather than from the root of the albums tree. So no surprises that it takes effectively the same amount of time in your case for the album at the top level that contains everything else.

Yes, it would help if you could instrument the code a bit; it's hard for me to get any useful timings here given how much faster it runs for me. The place to start for `Album::get` is https://github.com/LycheeOrg/Lychee-Laravel/blob/fix-316/app/Http/Controllers/AlbumController.php#L78. I would expect the bulk of the cost to be in the invocation of `get_albums` in  https://github.com/LycheeOrg/Lychee-Laravel/blob/fix-316/app/Http/Controllers/AlbumController.php#L146, but do confirm that first. This recursive function, along with most of the rest of the heavy lifting involved, is defined in https://github.com/LycheeOrg/Lychee-Laravel/blob/fix-316/app/ModelFunctions/AlbumFunctions.php. The easiest way to see all the SQL queries going on is to switch `APP_DEBUG` in `.env` to `true`; they'll then all end up in `storage/logs/laravel.log`.

**Comment by ildyria** [comment]
If you did install with composer with the dev mode then if you have `APP_DEBUG` to `true` Clockwork should also be working. https://github.com/itsgoingd/clockwork
You can either install the Firefox / Chrome extension or access it via : http://example.com/__clockwork

**Comment by matthijskooijman** [comment]
Thanks, clockwork works nicely (a bit slow, though).

It shows that (on my laptop, which is a bit easier to work with for me), a request takes 4500ms, of which 700ms is database (which I think means waiting for the database?). That suggests that the focus should probably not be the database queries.

Still, the number of queries is 941, which is about 4 queries per album. Looking at the queries themselves, this is one query to get the children, one query to get the album data, one query to get the photo's and one query to get the info of the user. Of these, I think the user queries should really be cached, and the children-query already returns the data that the single album query returns, so maybe the latter can also be optimized away.

![image](https://user-images.githubusercontent.com/194491/63230287-dd504e80-c20a-11e9-8d7c-f02a366634f2.png)

I've looked into the Albums::get case rather than Album::get, but have indeed confirmed that pretty much all the time is spent in `get_albums`: https://github.com/LycheeOrg/Lychee-Laravel/blob/b465213ad4581cf60316d6225761981930ba9674/app/ModelFunctions/AlbumFunctions.php#L258

I've also tried to use XDebug for profiling (since that looked like an easy way to dig into the details), but after I finally got that working (enabling it through the clockwork FF extension did not work for me) that showed meaningless data, probably due to [this bug](https://bugs.xdebug.org/view.php?id=77).

I spent a bit too much time on the profile, so I haven't been able to look more closely into `get_albums` (and I'm out of time now), but perhaps this is already useful input.

Until now, I have used `clock()->startEvent()` and `clock()->endEvent()` to time things, but AFAICS these require unique event names to work (e.g. if the code for the event is ran twice, such  as when I would add it to `get_albums()` which is called multiple times, clockwork gets confused rather than summing the time spent in the event). Do you know if there is another way to easily sum such time spans?

**Comment by kamil4** [comment]
By the sound of it, there may not be an obvious low-hanging fruit left?

It certainly feels like what we need here is an execution profile, but in the end you were unable to get a working one? @ildyria, do you have any suggestions? I remember that you use some 3rd party IDE for PHP development -- could it be of any help here in identifying the hot spots?

**Comment by kamil4** [comment]
@matthijskooijman I'm curious how much of the remaining slowdown is due to the album thumb calculation. Would you mind commenting out the two invocations of `gen_thumbs` on lines https://github.com/LycheeOrg/Lychee-Laravel/blob/fix-316/app/ModelFunctions/AlbumFunctions.php#L259 and https://github.com/LycheeOrg/Lychee-Laravel/blob/fix-316/app/ModelFunctions/AlbumFunctions.php#L460 and measuring again? Your most recent number on this branch was 22 seconds, correct?

Come to think of it, when you are done with the above experiment, try commenting out one line above the last one: https://github.com/LycheeOrg/Lychee-Laravel/blob/fix-316/app/ModelFunctions/AlbumFunctions.php#L459. This will, in principle, "break" Lychee but I expect it will also make it dramatically faster for you, with relatively little in terms of lost functionality. Specifically, the generation of album cover thumbs will be incomplete (pictures from subalbums will not be considered), and operations such as _Move_ or _Merge_ will only show top-level albums as possible destinations (again, no subalbums).

**Comment by kamil4** [comment]
@matthijskooijman There's one more experiment I'd like you to run: can you create a top-level album with ~1,000 pictures inside (all at the same level; no subalbums) and report on the performance of `Albums::get` vs `Album::get` with that album? That would need to be with a fresh database since your highly nested album tree will skew the results of `Albums::get`, of course. This experiment can be run on `master` or `fix-316` branch; it should make no difference.

**Comment by kamil4** [comment]
> ![image](https://user-images.githubusercontent.com/194491/63230287-dd504e80-c20a-11e9-8d7c-f02a366634f2.png)

I've been looking into optimizing some of these queries. So far I managed to get rid of the one originating from `ReadAccessFunctions::album`.

The remaining redundant `SELECT * FROM 'users'` (`Album.php:196`) originates from this line in the code:

```
$album['owner'] = $this->owner->username;
```

@ildyria, Is there some clever Laravel way of caching this query so that it doesn't repeat for every (sub)album? If not, I don't think it will be too much work to cache on our own (given how the code is structured, simply caching the last response should give us all of the benefit) but I thought I would ask first before possibly reinventing the wheel...

**Comment by ildyria** [comment]
https://laravel.com/docs/5.8/cache ? :)

**Comment by kamil4** [comment]
>     * https://laravel.com/docs/5.8/cache ? :)

I saw it last night and my first impression is that it seems way too heavy for this use case. It looks like the backend needs to be a file, or memcached, or... I just need to cache the result of a single query in RAM.

>     * https://code.tutsplus.com/tutorials/25-laravel-tips-and-tricks--pre-92818 see tip 5
>     * maybe worth trying : https://github.com/dwightwatson/rememberable

Yeah, that probably would've worked, but if I understand it correctly, this is an old feature that got removed, so now somebody implemented it back in a separate package? Unless we have another use case for this, that too looks like more effort than it's worth.

>     * https://laravel-news.com/laravel-model-caching

This one is giving me a headache :smile:. And it seems to use the Laravel Cache as well...

By the look of it, there is no ready-to-use, trivial solution to this problem so I'll probably do the caching on my own. We're talking about literally a single id to string mapping.

**Comment by kamil4** [comment]
I pushed my latest optimizations to PR #318. The number of queries we perform for `Albums::get` should be nearly halved, and `Album::get` should no longer be as slow as `Albums::get`. Even on my much faster server Lychee now feels considerably more snappy, especially on albums with many photos. I don't know if that will be enough to make Lychee usable on your slower hardware but I'm done optimizing the speed for now.

**Comment by matthijskooijman** [comment]
@kamil4, thanks for your work!

I just tried it on my server (with debug disabled) again, which is now down to 15 seconds (from 22s), so that's again a significant gain.

On my laptop (with debug enabled), it's down from 4500ms to around 3000ms. Queries are indeed down to 478 (from 941) and to 395ms (from 700ms).

Commenting out `gen_thumbs` (as suggested in https://github.com/LycheeOrg/Lychee-Laravel/issues/316#issuecomment-523268354, also the [newly introduced call/](https://github.com/LycheeOrg/Lychee-Laravel/blob/ee71d909f6aace78de46e6426c92bb32bbb24178/app/ModelFunctions/AlbumFunctions.php#L467)) reduces the time to around 1600ms. It also halves the query time and number (246 queries in 228ms).

Further looking inside `gen_thumbs` shows that the latter part (after the query, processing the query result) is negligable (adding a return after the query still shows around 3000ms). Putting the return before the query seems to also reduce the time to around 1600ms, which suggests that pretty much all time in `gen_thumbs()` is spent in the query (which is a lot more than the time Clockwork notes for db, but that might be just the query itself, maybe the bulk of the time is processing the query results on the Laravel model level).

On the server, I see slightly different behaviour. When I return before the query, it saves about half the time (from 15s to 8s), but when I return right after the query, I also go down from 15s to 11s, which suggests that the load is pretty well-divided between various parts of `gen_thumb()`.

> Come to think of it, when you are done with the above experiment, try commenting out one line above the last one: [/app/ModelFunctions/AlbumFunctions.php@`fix-316`#L459](https://github.com/LycheeOrg/Lychee-Laravel/blob/fix-316/app/ModelFunctions/AlbumFunctions.php#L459) [](/LycheeOrg/Lychee-Laravel/blob/HEAD@{2019-08-21T02:24:28Z}/app/ModelFunctions/AlbumFunctions.php#L459)

Limiting the recursion in Album::get in your commit reduced the runtime on my server to 5s for the top-level album (which contains just one album) and to 11s for the second-level album. Your above suggestion (commenting out `get_sub_albums()`) reduces `Album::get` to just over 1s for the top-level album and 3s for the second-level album, which would indeed make things somewhat usable (once the initial slow `Albums::get` is completed).


I do wonder: Why does `Albums::get` on startup actually return the whole tree of albums? I can imagine this speeds up further browsing, but whenever you open a subalbum, `Album::get` is called again, returning (mostly?) data that is already known from `Albums::get`? I can see that you need to iterate albums anyway (for thumbnails) and perhaps you need a list of albums for things like moving photos or navigating (though clicking open the album list at the top calls `Albums::get` *again*), but maybe that list does not need to include the thumbnails? I just tried calling `get_albums` with a recursion limit of 2 from `prepare_albums`, which is probably a hack, but (combined with the above suggested hack to not look at subalbums for thumbnails does make things somewhat usable for me, at first glance).

Thinking more on this, it seems that a lot of stuff can be a lot simpler if the thumbnail selection would be either explicit, or cached. Caching is of course tricky, since it is permission-dependent, but perhaps only using and caching the public thumbnail (e.g. the thumbnail selected from only public images) could be useful? Only if that does not exist could you then perhaps consider others?

**Comment by matthijskooijman** [comment]
I also tried optimizing the thumbnail query by adding an index to the `photos` table on the `album_id`, `star` and `takestamp` fields, but that did not seem to help. The EXPLAIN SQL output was unchanged (MySql did not use the new index) and in practice I might have seen about 1s improvement, though was likely just measurement noise.

**Comment by matthijskooijman** [comment]
@kamil4 thanks for clarifying. Summarizing, I guess that *if* we can cache or optimize the thumbnail generation, the rest of the code can probably be refactored to not require (deep) recursion and thus become a lot quicker (though maybe doing the recursion *just* to figure out thumbnails still has acceptable speed, especially if you can limit the data that is requested by the ORM by default).

One additional thought that I had was that maybe the database schema for the hierarchy could be adapted or supplemented. One particular problem with the way the hierarchy is stored now is that you cannot write a single SQL query that will list all descendants of a particular album. There are some alternative representations, I found a [concise summary](https://stackoverflow.com/questions/4048151/what-are-the-options-for-storing-hierarchical-data-in-a-relational-database) of them (there's also some better explanations around, but I haven't looked for them now).

Of these options, something like the [nested set model](https://en.wikipedia.org/wiki/Nested_set_model) might be useful, where each album would get a "left" and "right" number, and the numbers of subalbums always fit within the numbers of the parent. This could probably be added to the existing parent link (rather than replace it) and would allow getting all descendents of an album in a single query. You would still need to do permission checks, probably in PHP, and then do a second query to get the top-ranked photo in the set, but then it's only two queries rather than going through the entire hierarchy with one query for each album.

Alternatively, maybe things can also become a lot faster if you would go through the hierarchy to collect all subalbums, but not determine thumbnails for each subalbum and not even build a model object for each subalbum: Just collect the ids so you can use those in a thumbnail query later. This still has the recursion and one query per album, but saves a lot on the processing.


Then, even with the above improvements, there is the case of generating a full hierarchy of albums *with* thumbnails (as is needed for the album selection dropdown). None of my above suggestions will really help for that case, AFAICS, so that does really need thumbnail caching.

**Comment by ildyria** [comment]
> We do already cache one piece of information in the albums, and that's `min_takestamp` and `max_takestamp`. We calculate it based on all the content inside, ignoring the permissions, because the decision was made that the information leak is minimal and acceptable. That wouldn't be the case with thumbnails; we must not allow private pictures to leak. Doing the calculation just for the public mode wouldn't help if the owner of the pictures logs on, or the admin, and that's probably common for most people's albums. My feeling is that if we were to do it, we should calculate two sets: one for the public mode and another for the album owner (which would include the admin). That would cover virtually all important scenarios (there are some edge cases of subalbum sharing with individual users but those could probably be ignored). I recently revamped the code to calculate the `takestamp` columns in an optimal fashion, using the smallest possible number of updates (we used to recalculate everything on album updates); we could probably reuse some of it for thumbs as well...

Creating a pivot table containing `album_user_photo_id`, `album_id`, `user_id`, `photo_id` would indeed remove the recursions etc and enable caching. However... When adding/moving/deleting/duplicating a picture from/to an album the computational amount will be a nightmare as we need to regenerate the thumbs for all the users with whom the albums have been shared (in addition to public & admin).

**Comment by d7415** [comment]
> Of these options, something like the nested set model might be useful...

> Creating a pivot table containing album_user_photo_id, album_id, user_id, photo_id would indeed remove the recursions etc and enable caching. However... When adding/moving/deleting/duplicating a picture from/to an album the computational amount will be a nightmare as we need to regenerate the thumbs for all the users with whom the albums have been shared (in addition to public & admin).

Yeah, these are both just shifting the problem from expensive reads (which appear to only be a problem with extremely large albums and/or extremely slow servers) to very expensive writes. The linked article even says as much about nested sets.

Optimising is good, but dedicating lots of work into making one part faster at the expense of another isn't, especially if it introduces complexity.

@kamil4's solution may be a good compromise, but I'm not sure how the cost/benefit would work out. Alternatively, we could just store the public one, and if it's empty run the generator? I don't know whether this would help...

**Comment by kamil4** [comment]
FYI, recent commits, specifically fb2d66c604c371140c10c8ea82fcb46927f66e6e and its front-end counterpart, reduced the recursion limit of `get_albums()` by 1 on `Album::get` calls, so those should become less expensive. The cost of `Albums::get` should be the same though.

**Comment by zhanglix** [comment]
I encountered the same issue. I migrate lots of photos from 'shotwell' to lychee yesterday and got about 500+ albums because 'shotwell' grouped all photos by date taken.  And it is very slow whenever need to load albums list .

I spend about  half a day to merge all those 'daily albums' into 'monthly' one to reduce the amount of albums to less then 100.  Now the loading time is about 2-3s, still slow but tolorable.

My question is it possible to sort all photos into 'logical albums' of years and months, and generate albums list and their thumbs into static file or cache automatically so that load albums lists can be done in several milliseconds  instead of seconds?

**Comment by ildyria** [comment]
> My question is it possible to sort all photos into 'logical albums' of years and months, and generate albums list and their thumbs into static file or cache automatically so that load albums lists can be done in several milliseconds instead of seconds?

Hi. Can you open a new issue with your question ?

> I encountered the same issue. I migrate lots of photos from 'shotwell' to lychee yesterday and got about 500+ albums because 'shotwell' grouped all photos by date taken. And it is very slow whenever need to load albums list .

This really highlight where the problem comes from.
You may be one of the biggest "album" user so far. :)

**Comment by zhanglix** [comment]
> FYI, recent commits, specifically [fb2d66c](https://github.com/LycheeOrg/Lychee/commit/fb2d66c604c371140c10c8ea82fcb46927f66e6e) and its front-end counterpart, reduced the recursion limit of `get_albums()` by 1 on `Album::get` calls, so those should become less expensive. The cost of `Albums::get` should be the same though.

That sound great!  Since I am using  the  docker image from https://hub.docker.com/r/lycheeorg/lychee-laravel on a raspberry pi board , could you publish a nightly docker image for linux/arm/v7 arch?

**Comment by ildyria** [comment]
> That sound great! Since I am using the docker image from https://hub.docker.com/r/lycheeorg/lychee-laravel on a raspberry pi board , could you publish a nightly docker image for linux/arm/v7 arch?

@d7415 ? :laughing:

**Comment by d7415** [comment]
We already publish a nightly under the `:dev` tag.

**Comment by zhanglix** [comment]
> We already publish a nightly under the `:dev` tag.

Cool! 👍  I will try it once I sync all my photos 😸

**Comment by zhanglix** [comment]
> > My question is it possible to sort all photos into 'logical albums' of years and months, and generate albums list and their thumbs into static file or cache automatically so that load albums lists can be done in several milliseconds instead of seconds?
> 
> Hi. Can you open a new issue with your question ?
> 

I opened a new issue. #549

**Comment by zhanglix** [comment]
I tried the new docker image with tag :dev, but can't make it work. 

I stop and renamed the old.container. and started a new one with the new image. And all I can.got after login is like this:
![Screen Shot 2020-05-13 at 1 01 02 AM](https://user-images.githubusercontent.com/8109462/81724143-79676780-94b6-11ea-8ac0-27b940aacf4c.png)

Are there any additional steps need to do to upgrade Lychee?

**Comment by d7415** [comment]
> I tried the new docker image with tag :dev, but can't make it work.

Please open a new issue, using the issue template.

**Comment by zhanglix** [comment]
I manage to move all my photos into one big directory and started syncing. And I found this is what I really like -- all photos are sorted by shot time in one photo stream. And the performance of album::get is not important any more. :P

**Comment by ildyria** [comment]
hahaha :D

**Comment by zhanglix** [comment]
I encountered another extreme case :) . Since there are about 12000photos/videos in the "NULL" albums, Album::get takes about 28s to finish and returned about 13M bytes.
<img width="387" alt="Screen Shot 2020-05-14 at 9 24 48 PM" src="https://user-images.githubusercontent.com/8109462/81939778-5ad5ad80-9629-11ea-8283-b0ad12f99d13.png">
<img width="387" alt="Screen Shot 2020-05-14 at 9 25 30 PM" src="https://user-images.githubusercontent.com/8109462/81939841-72149b00-9629-11ea-8014-d5abfd7ed61f.png">

Is it possible to just get a subset of  photo records  per Album:get (eg 100)?  Further call to Album::get will be triggered when user scrolls down to  the end of current page?

**Comment by ildyria** [comment]
:scream: 

This will require some rewrite and unfortunately that is not a top priority at the moment. :/
This will likely be solved with the new front end that I have in mind.

**Comment by zhanglix** [comment]
> This will require some rewrite and unfortunately that is not a top priority at the moment. :/

OK, I can sort those photos into several albums as a workaround. 
Anyway, I think Lychee is the best self host photo management service I can find. You guys have done a  really great job! 👍 👍 👍  

> This will likely be solved with the new front end that I have in mind.

If you are planning a new front end,  I  wish you will lend  some ideas from iCloud photo library. such as more smart albums based on different criteria like types( video, selfies, screenshot), take stamp range, location, faces recognized.  :)

**Comment by ildyria** [comment]
@zhanglix can you check with the latest version with respect to the speed. There should be a significant improvement in speed as we changed the tree structure.
